### PR TITLE
Twitter processor

### LINKED ISSUE #17: Twitter processor
(Part of #3)

Porting the existing twitter processor.

- [x] `social-media` pipeline generator
  - [x] example `source-spec.yaml` format for pipeline generator
- [x] `twitter` processor
  - [x] `source` column
  - [x] `followers`
  - [x] `mentions`
  - [x] `interactions`
- [x] sql datastore class 
- [x] tests
- [x] document new env vars

### PR DESCRIPTION
**Pull request by brew**
This pull request fixes #17 and is connected to #3.

* [x] I've added tests to cover the proposed changes

Changes proposed in this pull request:

- Fleshes out the socialmedia pipeline
- Adds a twitter processor
  - adds `followers`, `mentions` and `interactions`
- Adds an `SQLDatastore` interface

Processor retrieves two types of metrics, 'lifetime' metrics, such as `followers`, that are retrieved from the api, and used as-is without processing, and 'cumulative' metrics, such as `mentions` and `interactions` that are retrieved for a specific time-frame, then added to the latest stored result.

### CODE REVIEW

File: datapackage_pipelines_measure/pipeline_steps/social_media.py [hunk 1]
```diff
@@ -1,23 +1,74 @@
...
+    for entity in config['twitter']['entities']:
+        steps.append(('measure.add_twitter_resource', {
+            'entity': entity,
+            'project_id': project_id
+        }))
+
+    steps.append(('concatenate', {
+        'target': {
+            'name': 'social-media',
+            'path': 'data/social-media.json'},
```
Comment (akariv) [inline datapackage_pipelines_measure/pipeline_steps/social_media.py]: Why do you add a path here? Why the json extension?

File: datapackage_pipelines_measure/processors/add_twitter_resource.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+
+parameters, datapackage, res_iter = ingest()
+
+entity = parameters['entity']
+project_id = parameters['project_id']
+safe_entity = _get_safe_entity(entity)
+resource = {
+    'name': safe_entity,
+    'path': 'data/{}.json'.format(safe_entity)
```
Comment (akariv) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: Why the json?

File: datapackage_pipelines_measure/processors/add_twitter_resource.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    'interactions': interactions
+}
+
+resource['schema'] = {
+    'fields': [{'name': h, 'type': 'string'} for h in resource_content.keys()]}
+
+datapackage['resources'].append(resource)
+
+# Make a single-item generator from the resource_content
+resource_content = itertools.chain([resource_content])
```
Comment (akariv) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: that's not really needed.

File: datapackage_pipelines_measure/datastore/sql.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+from sqlalchemy.ext.automap import automap_base
```
Comment (akariv) [inline datapackage_pipelines_measure/datastore/sql.py]: See later comment on why I think you don't need this.

File: datapackage_pipelines_measure/processors/add_twitter_resource.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+# Get last run entity values
+filter_object = {
+    'entity': entity,
+    'entity_type': entity_type,
+    'source': 'twitter',
+    'project_id': project_id
+}
+datastore = get_datastore()
+entity_last_run = datastore.get_latest_from_table(filter_object,
```
Comment (akariv) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: Not sure where this comment belongs so I'm putting it here.

I think that a simpler solution would be to always query 3 times for the past three days (i.e. today, yesterday & the day before that).
Then you generate 3 separate records for each of these query results.

When writing to the DB, you use the timestamp (and probably entity, entity_type and source) as key for updating the DB.

This way you don't need to store any permanent state (which will be deleted once the container is re-deployed).
Also, I think that conceptually aggregation can (and should) be done separately, and not in the raw collector - either in a separate pipeline or in a simple SQL query.

File: datapackage_pipelines_measure/processors/add_twitter_resource.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        start_date.strftime(TWITTER_API_DATE_RANGE_FORMAT),
+        end_date.strftime(TWITTER_API_DATE_RANGE_FORMAT)
+    )
+    interactions = _get_interactions_for_entity_from_source(
+        entity,
+        start_date.strftime(TWITTER_API_DATE_RANGE_FORMAT),
+        end_date.strftime(TWITTER_API_DATE_RANGE_FORMAT)
+    )
+
+resource_content = {
```
Comment (akariv) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: no timestamp in record? 
I think that we should collect the raw data here and do aggregations later.

File: datapackage_pipelines_measure/pipeline_steps/social_media.py (hunk 1 above)
Comment (brew) [inline datapackage_pipelines_measure/pipeline_steps/social_media.py]: The path is useful if I want to `dump.to_path` in development. I have a patch lined up that will dump the latest to path, only there's a DEVELOPMENT flag in the env.

The `.json` extension is a mistake, and interestingly, is ignored when the resource file is dumped, and a `.csv` extension is used instead.

File: datapackage_pipelines_measure/processors/add_twitter_resource.py (hunk 2 above)
Comment (brew) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: Again, this should be a `.csv`. Though, most of the time this path is never used, as the resources are concatenated later in the pipeline.

File: datapackage_pipelines_measure/processors/add_twitter_resource.py (hunk 6 above)
Comment (brew) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: Timestamp is added in a later processor in the pipeline.

File: datapackage_pipelines_measure/processors/add_twitter_resource.py (hunk 3 above)
Comment (brew) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: If I don't do this, I get an error from `concatenate`:

```
INFO    :Main                            :concatenate: /Users/brew/virtualenvs/measure/lib/python3.6/site-packages/jsontableschema/model.py:48: UserWarning: Class models.SchemaModel is deprecated [v0.7-v1)
INFO    :Main                            :concatenate:   warnings.warn(message, UserWarning)
INFO    :Main                            :concatenate: Traceback (most recent call last):
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/specs/../lib/concatenate.py", line 119, in <module>
INFO    :Main                            :concatenate:     spew(datapackage, new_resource_iterator(resource_iterator))
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/wrapper/wrapper.py", line 59, in spew
INFO    :Main                            :concatenate:     for rec in res:
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/specs/../lib/concatenate.py", line 99, in concatenator
INFO    :Main                            :concatenate:     in row.items()
INFO    :Main                            :concatenate: AttributeError: 'str' object has no attribute 'items'
```

File: datapackage_pipelines_measure/processors/add_twitter_resource.py (hunk 5 above)
Comment (brew) [inline datapackage_pipelines_measure/processors/add_twitter_resource.py]: After some discussion offline, we decided not to aggregate data in the processor, instead leaving aggregation as a task for the reporting layer (Redash/SQL). This should simplify collection.

**Comment by brew** [comment]
@akariv I've refactored the twitter processor so it no longer aggregates locally. 

I decided to keep the timestamp as it is, as it represents the datetime the data was collected, and I'd like to differentiate between the time collected (metadata) and the date for the data we're interested (which itself is data). So I've added a `date` column to represent the data date.

**Pull request closed without merging**